### PR TITLE
docs: add sphinx setup

### LINKED ISSUE #235: Sphinx docs? Read the docs?

### PR DESCRIPTION
**Pull request by austin3dickey**
Closes #235.

This PR adds a `docs/` folder containing a fairly minimal Sphinx configuration that can be used for a GitHub Pages website. I think this website will add value because of its:
- search bar and sidebar
- explorability (everything is right next to each other and self-linked)
- cleanliness (no directory structure or code files to sift through)

I added the READMEs and docstrings to the website. Of course, as we add more pages and flesh out the docstrings and type hints, they will show up on the website too.

This PR also includes a new GitHub Actions workflow to deploy the GitHub Pages site based off the current `main` branch of the repo. It is triggered whenever we release a package to PyPI, but also can be run manually if we wish.

I validated that the new workflow is successful on my fork. Check out the site! https://austin3dickey.github.io/conbench/

### CODE REVIEW

**Comment by austin3dickey** [comment]
This is ready to go; check out an example of the site in the PR description.

File: README.md [hunk 1]
```diff
@@ -47,18 +49,6 @@ repository, and the results are hosted on the
...
-## Index
-
-* [Developer environment](https://github.com/conbench/conbench#developer-environment)
-* [Configuring the server](https://github.com/conbench/conbench#configuring-the-server)
-* [Creating accounts](https://github.com/conbench/conbench#creating-accounts)
-* [Authoring benchmarks](https://github.com/conbench/conbench#authoring-benchmarks)
-  * [Simple benchmarks](https://github.com/conbench/conbench#example-simple-benchmarks)
-  * [External benchmarks](https://github.com/conbench/conbench#example-external-benchmarks)
-  * [Case benchmarks](https://github.com/conbench/conbench#example-case-benchmarks)
-  * [R benchmarks](https://github.com/conbench/conbench#example-r-benchmarks)
```
Comment (austin3dickey) [inline README.md:59]: These were already out-of-date, and the new site has everything on the left sidebar.

**Review by jgehrcke — COMMENTED** [review]
@austin3dickey thank you for this very specific proposal.

To be honest, after our brief discussion in #235 I did not expect the focus of this PR :tropical_fish:!

There, you stated last that

> (1) is already available via clicking around the repository and viewing GitHub Markdown.

My opinion is that I do not think that (1) is available. The content and structure should be taken to the next level. That would be for now in my opinion the most important documentation area to invest love and time into.

I think we should prioritize building content and structure first with Markdown files in `docs/`. And later potentially build a docs page from that, as sugar on top. And even later build a dedicated site for API docs.

For making content/structure work as friction-less as possible I find it important that initially as _author and reviewer_ we only need to look at GitHub rendered Markdown when we review each other's documentation PRs. That is one reason for why I expressed in #235 as my wish:

> The general documentation (1) should render nicely in the GitHub UI straight from Markdown files.

I am not sure about your opinion on that, and how this relates to this PR here. (would love to see discussion!).

When optimizing CI and review/authoring process for writing _content_ then one often talks about stuff like [one-sentence-per-line](https://nick.groenen.me/notes/one-sentence-per-line/).

After our short exchange of opinions in #235 I am surprised to read:

> Add a file to `docs/pages/`. It can be RST or Markdown.

It almost feels like an oversight, but is it? Would have loved to see a little bit of a 'motivation/justification' for this, also more direct responses to the rather strong opinions expressed before in #235.

Here, we add Sphinx and the main focus of this PR seems to be API docs. I love Sphinx: first time usage for me was 2008 to build [this website](https://gehrcke.de/awsac/) that I am a bit embarrassed about as of today, have [reported bugs](https://github.com/sphinx-doc/sphinx/issues/37#issuecomment-68523721) over the years, have carefully written [API reference docs](https://gehrcke.de/gipc/api.html) using RST, and I customized and contributed to sphinx-rtd-theme. This is just to hopefully convincingly describe that I do have feedback on the details here, and generally I have opinions that I think start to matter once we are buying into the Sphinx ecosystem. I want to have _passion_ for us doing sphinxy things, I'd love to add my opinions, but I feel a little bit overrun by this patch. As of my view on priorities I am just not sure if these days is already a good time for talking about such details at all. 

Generally this feels like we should collaborate closer and find better alignment on priorities and problem statements! We will get there. I don't want to block this pull request at all. But I'd also love to be able to re-think some of what's in here at a later point in time.

File: benchadapt/python/benchadapt/result.py [hunk 2]
```diff
@@ -114,7 +116,7 @@ class BenchmarkResult:
     - ``run_reason``
     - ``github``
     - ``machine_info``
-    - ``cluster_info
+    - ``cluster_info``
```
Comment (alistaire47) [inline benchadapt/python/benchadapt/result.py:119]: oops thank you

**Review by alistaire47 — COMMENTED** [review]
Interesting, this really shows how lax we are at writing docstrings. Especially for the client libraries I think it'd be worth it to do a quick pass and document methods and params and such.

For the main webapp, I'm less sure exactly what the right way to document everything is. I would personally find it very useful to have lots of docstrings and such, but the reality is the code there is only intended for use in the context of the package, so the idea of documenting things that are "exported" or "exposed to users" (scare quotes because in Python these things are informal) doesn't really help.

**Comment by austin3dickey** [comment]
@jgehrcke To provide further background, I decided to create this PR today because
- In our team's longer-term goals I found (and agreed with) this statement from @jonkeane:  "At the very least we need to host *some* docs so we can put some of the documentation we have (and will encourage us to keep adding more)."
- I've set up something like this on dozens of projects before. Each time, I thought it provided some value. (See the PR description for the 3 pieces of value I think this solution adds.) This is definitely not to say it's the best solution! Just that I knew I could knock it out in less than a day and hopefully quickly turn on a cool, high-visibility feature.
- I wanted to propose a concrete, coded, merge-ready alternative to linking people to Markdown directories in the GitHub repo. That way we can have a productive discussion over the pros and cons. :) 

I definitely did not mean to make you feel "overrun," or at a point where you feel you didn't have an input, so I apologize for that!

To clarify: when I said 
>"Add a file to `docs/pages/`. It can be RST or Markdown."

I only meant that in the technical sense. With this setup, either format will be rendered correctly. I'm happy to amend that sentence to only recommend GitHub Flavored Markdown, so no one feels the temptation to write in RST. That is still completely in accordance with this PR's setup. Does that address your concern?

>When optimizing CI and review/authoring process for writing content then one often talks about stuff like [one-sentence-per-line](https://nick.groenen.me/notes/one-sentence-per-line/).

That is also in accordance with this PR's setup! I think it's outside the scope of this PR to change everything existing today to one-sentence-per-line, though.

> I am just not sure if these days is already a good time for talking about [sphinx] details at all ... I'd also love to be able to re-think some of what's in here at a later point in time.

Of course, this is completely fine! I appreciate that you also have lots of sphinx experience and am excited to learn some new tips. But I still think this PR adds incremental value at a low cost. I don't think we need to have it perfect on the first try. But if you'd rather we hold off until we have more time to make it better, we can decline this PR and come back to it later.

**Comment by jgehrcke** [comment]
> Each time, I thought it provided some value.

Yes it provides value! We just have to be mindful of these ideas:
- Certain type of work and process has a bit of a drag/slowdown effect in the mid/long-term.
- There is an infinite amount of (valuable) work that can be done, but it also takes infinitely long. Setting micro goals and priorities and spending time on refining problem statements (before entering solution space) is important for reaching more ambitious, bigger goals.

That's one perspective that I am coming from.

> I definitely did not mean to make you feel "overrun," or at a point where you feel you didn't have an input, so I apologize for that!

Thank you, and I believe that best intentions are here.

It helps so much to find alignment on goals and priorities before expanding maintenance surface and complexities. It's OK to spend a little longer on the alignment part, and then save a bunch of time and energy later on.

It's a really important task to remove ambiguity from "we need to host some docs so we can put some of the documentation we have", to translate this into into good problem statements / goal statements. In my opinion we had nicely started this process in #235; but then entered solution space a little prematurely. I then saw a dissonance between the solution space and the not-yet-fleshed-out priorities. (I know I don't follow this order of things with perfection either, and want to be held accountable by my peers!).

Below I can hopefully clarify how much in this here has to do with communciation around setting goals.

I imagine a scenario where #235 would have resulted in the goal "let's have _sphinx-rendered_ and polished RST-based library docs covering the API surface of library A, and library B". I would have disagreed-and-committed in there. Then this PR would cover the "_sphinx-rendered_" part of that. The library documentation content work would be a rather longish stream of subsequent work that I am a little skeptical about as of today.

In contrast, the discussion in #235 started to go into the direction of "let's have github markdown-rendered docs for Conbench in general" which is a rather different goal. I am not saying that we in fact _did_ achieve alignment in there.

> alternative to linking people to Markdown directories in the GitHub repo.

An efficient way is to start having this directory, and to create _great content_ in there. And to actually link people there. It gets us so so so far! So many big, serious engineering projects do precisely this. And of course, once some decent content is there, one can add this alternative. With specific goals in mind; and better decision-making ability because there's content there already.

> I only meant that in the technical sense. With this setup, either format will be rendered correctly

Yes. But flexibility in this case is not a feature; and this statement indeed sounds like we are happy with having a mixture. My default perspective on ambiguity is that it has cost associated with it. Removing ambiguity and setting clear goals and expectations helps. It's helpful to say "docs must be in markdown, and that markdown is linted by CI".

> I think it's outside the scope of this PR to change everything existing today to one-sentence-per-line, though.

That's a lovely discussion to have, because I argue that if one focuses on _content first_ then for efficient reviews and great diffs, this is a great candidate to introduce and settle on early!

Thanks for allowing for us to have this exchange; as we can see it's all about more global goals and priorities, and not about the details in this PR.

**Comment by jonkeane** [comment]
I'm supportive of merging this as is, but want to do so with a forward-looking discussion about things that we will (and won't) prioritize. 

I'm especially cognizant of what JP summarized in (and then continued on in detail later, but for clarity of quoting I'm leaving that out here!):

> * Certain type of work and process has a bit of a drag/slowdown effect in the mid/long-term.
> * There is an infinite amount of (valuable) work that can be done, but it also takes infinitely long. Setting micro goals and priorities and spending time on refining problem statements (before entering solution space) is important for reaching more ambitious, bigger goals. 

Generally, I think we should proceed with merging this, but following this focus on the broad documentation we all agree is lacking right now, and explicitly defer things like: refining sphinx theming and other presentation elements, improving our broad docstring coverage for old code, and if publishing this to html ends up getting in the way of forward progress (compared to using "just" github rendered markdown) we will reconsider and are open to pulling it back out (I don't think it will — and the ways it can do that are slightly pernicious if it does, like not starting a PR that should be a simple code change, but will impact lots of docs — but good to call this out and all be on board before we get into that situation).

We can (and will!) concentrate on those things we're deferring now (they are valuable!), but we should use this as an opportunity to push on the most important and lacking docs (the long form docs explaining how conbench works broadly and how the pieces fit together). Generally, those should be (could we agree must be?) markdown, and ideally render in GitHub just as well as they do on the site (this keeps us concentrated on the content, and let's us pull back to just markdown or transition to a different docs generator if we want to when we get to the refining stages).  

Then once those general docs are put together, we can determine the relative priority of the other bits of cleanup (in comparison both to other features we are developing and other kinds of docs improvements) as we get to them. One concrete example of this is: what level(s) of documentation we give to our individual libraries will and should vary greatly, `benchconnect` should generally only be used by us, so can have relatively spartan documentation, whereas things like `benchadapt` will have outside consumers using it, so will need more details. Likely we will have to defer lots of work here in favor of other work in the short to mid term (and that's ok!). But I want to make sure we're all on the same page about that.

File: README.md [hunk 3]
```diff
@@ -205,10 +195,20 @@ The following commands are not guaranteed to work as documented, but provide val
 Kick off a new run of the "Build and upload a package to PyPI" workflow on the [Actions
 page](https://github.com/conbench/conbench/actions).
 
+### To add new documentation pages
+
+To add a new page to our GitHub Pages-hosted documentation:
+
+1. Add a Markdown file to `docs/pages/`.
+2. In the toctree in the `docs/index.rst` file, add `pages/your_new_page`, where
+   `your_new_page` is your new filename without the `.md` file suffix.
```
Comment (austin3dickey) [inline README.md:204]: I clarified this section of the README!

**Comment by austin3dickey** [comment]
Tests are failing due to https://github.com/conbench/conbench/issues/758

**Pull request merged**